sudoku/lib/sudoku/game.ex

```diff
@@ -0,0 +1,151 @@
+defmodule Sudoku.Game do
+  # To-Do:
+  # Generate board function
+  # Check board function
+  #   contains: check column, check row, check group
+
+  # put move
+  # print board
+
+  # Useful lower-level commands so far:
+  # board = Sudoku.Game.new_game()
+  # board = Sudoku.Game.add_move(board, "A9", 2)
+  # Sudoku.Game.print_board(board)
+
+  def new_game() do
+    # To-do:
+    instantiate_board()
+    |> generate_puzzle()
+  end
+
+  # The gist of the general algorithm:
+  #   1. Assumption: Given the rules of sudoku, given a 9x9 board that can be broken up
+  #      into 3x3 segments, the first 3 diagonal 3x3 segments can be randomly generated with numbers 1-9
+  #      with no collisions.
+  #   2. After the first 3 3x3 diagonal segments are generated, randomly populate each other 3x3 segment
+  #      in the board. At each individual square, check to see if there are column collisions or row collisions.
+  #   3. Remove elements to complete the "puzzle". There are smart ways to do this and dumb ways to do this.
+
+  # Next steps:
+  # 1. Add validation to the "add move" that can't overwrite original generated values
+  # 2. Perhaps create a game struct that contains the board.
+  # 3. Add concurrency
+  # 4. Create GUI for webapp
+
+  @rows   'ABCDEFGHI'
+  @columns '123456789'
+
+  def generate_puzzle(board) do
+    for row <- @rows do
+       for col <- @columns do
+        r = List.to_string([row])
+        c = List.to_string([col])
+        # IO.puts(r)
+        # IO.puts(c)
+        board = add_move(board, Enum.join([r, c]), Enum.random(1..9))
+       end
+    end
+  end
+
+  def check_column(board, val) do
+
+  end
+
+  def check_row(board, val) do
+
+  end
+
+  def check_group(board, val) do
+
+  end
+
+  # If every value is non-zero, then check for a win.
+  def check_win(board) do
+
+  end
+
+  # To-do with add move:
+  # once check_group, check_row, and check_column is implemented, integrate this into validity checking
+  # for add_move.
+
+  def add_move(board, coord, move) do
+    valid_coord = coord
+    |> String.match?(~r/[a-iA-I][1-9]$/)
+
+    valid_move = move
+    |> Integer.to_string
+    |> String.match?(~r/[1-9]$/)
+
+    board
+    |> try_move(coord, move, valid_coord, valid_move)
+  end
+
+  # Regex: // ~r/[A-I][1-9]/
+  def try_move(board, coord, move, true, true) do
+    [r, c] = coord
+    |> String.capitalize()
+    |> String.codepoints()
+    row = r |> String.to_atom
+    Kernel.put_in(board[row][c], move)
+  end
+
+  def try_move(_board, _string, _move, false, true) do
+    IO.puts("Invalid coordinate placement!")
+    :invalid_coord
+  end
+
+  def try_move(_board, _string, _move, true, false) do
+    IO.puts("Invalid value!")
+    :invalid_value
+  end
+
+  def try_move(_board, _string, _move, _valid_coord, _valid_move) do
+    IO.puts("Invalid coordinate placement and value!")
+    :invalid_both
+  end
+
+  def instantiate_board() do
+    _board = %{
```

Comment:
Given that data structures are immutable, you should also be able to do (untested)

~~~ elixir
empty_columns = %{"1" => 0, "2" => 0, "3" => 0,   "4" => 0, "5" => 0, "6" => 0,   "7" => 0, "8" => 0, "9" => 0}
board = for row <- [ A, B, C, D, E, F, G, H, I ], do: { row, empty_columns }, into: %{}
~~~